### PR TITLE
adds: prebuild (blue/green) style deployment script

### LINKED ISSUE #1130: Blue / Green Deployments in Staging Environment
**What would you like to be added**:
- Blue / Green style deployment of Telescope Services

**Why would you like this to be added**:
Ensure Staging -and by extension once implemented in Production, is never down when new commits / builds are released.

### PR DESCRIPTION
**Pull request by raygervais**
## Issue This PR Addresses
fixes #1130 

## Type of Change

- [X] **Bugfix**: Change which fixes an issue
- [X] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

Adds pre-building of Docker containers prior to spinning down environment, also doesn't destroy services so that we can switch within seconds to either environment. `Dockerfile` is updated to now leverage multiple containers for each dependency (frontend, backend) so that work between the two doesn't trigger a complete npm install chain. It's the small optimizations. 

*TODO*: Integrate into `deploy.sh` logic

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

## Script Output Example
```bash
Successfully built 7577b012e84b
Successfully tagged green_telescope:latest
Waiting...
Stopping blue Container
Stopping telescope_production ... done
Stopping blue_redis_1         ... done
Stopping blue_elasticsearch_1 ... done
Removing orphan container "blue_certbot_1"
Removing orphan container "blue_logrotate_1"
Removing orphan container "blue_nginx_1"
Removing telescope_production ... done
Removing blue_redis_1         ... done
Removing blue_elasticsearch_1 ... done
Removing network blue_default
Starting green Container
Creating network "green_default" with the default driver
Creating green_redis_1         ... done
Creating green_elasticsearch_1 ... done
Creating green_logrotate_1     ... done
Creating green_certbot_1       ... done
Creating telescope_production  ... done
Creating green_nginx_1         ... done
bash

### CODE REVIEW

**Comment by humphd** [comment]
I'm really excited to see this happening

File: tools/autodeployment/deploy.sh [hunk 1]
```diff
@@ -1,30 +1,24 @@
-#!/bin/bash
-set -e
-set -u
-set -x
+#!/bin/bash -e -u -x
```
Comment (humphd) [inline tools/autodeployment/deploy.sh:1]: Interesting, I haven't seen this done on the shebang before.

**Review by humphd — APPROVED** [review]
Untested by me, but code looks sound.  Get @manekenpix to approve the actual function of it.

File: tools/autodeployment/deploy.sh (hunk 1 above)
Comment (raygervais) [reply tools/autodeployment/deploy.sh:1]: What's interested is I've never seen it done the inverse way haha.

**Review by manekenpix — APPROVED** [review]

File: tools/autodeployment/deploy.sh (hunk 1 above)
Comment (raygervais) [reply tools/autodeployment/deploy.sh:1]: Reverting since the version of BASH on Fedora doesn't enjoy this syntax.

**Review by manekenpix — APPROVED** [review]

**Review by manekenpix — APPROVED** [review]

**Review by humphd — APPROVED** [review]

**Review by manekenpix — APPROVED** [review]

**Comment by humphd** [comment]
Thanks for fixing that markdown issue for me.  Apologies for breaking the tree earlier.

**Pull request merged**